File: services/api/models/knowledge.go
package models

import (
	"time"

	"github.com/oklog/ulid/v2"
	"gorm.io/gorm"
)

// Knowledge represents a knowledge base entry
type Knowledge struct {
	ID          string     `json:"id" gorm:"primaryKey"`
	Title       string     `json:"title" gorm:"not null"`
	Content     string     `json:"content" gorm:"type:text"`
	CategoryID  string     `json:"category_id" gorm:"index"`
	AuthorID    string     `json:"author_id" gorm:"index"`
	Tags        []Tag      `json:"tags" gorm:"many2many:knowledge_tags;"`
	Status      string     `json:"status" gorm:"default:'draft'"` // draft, published, archived
	Version     int        `json:"version" gorm:"default:1"`
	ParentID    *string    `json:"parent_id" gorm:"index"` // For version history
	ViewCount   int        `json:"view_count" gorm:"default:0"`
	Source      string     `json:"source"`    // obsidian, zotero, forum, manual
	SourceID    string     `json:"source_id"` // Original ID from source system
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	PublishedAt *time.Time `json:"published_at"`
}

// BeforeCreate generates ULID before inserting
func (k *Knowledge) BeforeCreate(tx *gorm.DB) error {
	if k.ID == "" {
		k.ID = ulid.Make().String()
	}
	return nil
}

// TableName specifies the table name
func (Knowledge) TableName() string {
	return "knowledge"
}

// Category represents a knowledge category
type Category struct {
	ID          string    `json:"id" gorm:"primaryKey"`
	Name        string    `json:"name" gorm:"not null"`
	Description string    `json:"description"`
	ParentID    *string   `json:"parent_id" gorm:"index"`
	Level       int       `json:"level" gorm:"default:1"` // 1, 2, 3 for tree depth
	SortOrder   int       `json:"sort_order" gorm:"default:0"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// BeforeCreate generates ULID before inserting
func (c *Category) BeforeCreate(tx *gorm.DB) error {
	if c.ID == "" {
		c.ID = ulid.Make().String()
	}
	return nil
}

// TableName specifies the table name
func (Category) TableName() string {
	return "categories"
}

// Tag represents a knowledge tag
type Tag struct {
	ID        string    `json:"id" gorm:"primaryKey"`
	Name      string    `json:"name" gorm:"uniqueIndex;not null"`
	Color     string    `json:"color" gorm:"default:'#1890ff'"`
	Count     int       `json:"count" gorm:"default:0"` // Usage count
	CreatedAt time.Time `json:"created_at"`
}

// BeforeCreate generates ULID before inserting
func (t *Tag) BeforeCreate(tx *gorm.DB) error {
	if t.ID == "" {
		t.ID = ulid.Make().String()
	}
	return nil
}

// TableName specifies the table name
func (Tag) TableName() string {
	return "tags"
}

// ZoteroItem represents a Zotero library item
type ZoteroItem struct {
	ID          string    `json:"id" gorm:"primaryKey"`
	ZoteroKey   string    `json:"zotero_key" gorm:"uniqueIndex"`
	Title       string    `json:"title"`
	ItemType    string    `json:"item_type"`                // book, journalArticle, etc.
	Authors     string    `json:"authors" gorm:"type:text"` // JSON array
	Abstract    string    `json:"abstract" gorm:"type:text"`
	Publication string    `json:"publication"`
	Volume      string    `json:"volume"`
	Issue       string    `json:"issue"`
	Pages       string    `json:"pages"`
	Date        string    `json:"date"`
	DOI         string    `json:"doi"`
	URL         string    `json:"url"`
	PDFPath     string    `json:"pdf_path"`
	Tags        string    `json:"tags" gorm:"type:text"` // JSON array
	SyncedAt    time.Time `json:"synced_at"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName specifies the table name
func (ZoteroItem) TableName() string {
	return "zotero_items"
}

// KnowledgeReview represents a knowledge review/approval record
type KnowledgeReview struct {
	ID          string    `json:"id" gorm:"primaryKey"`
	KnowledgeID string    `json:"knowledge_id" gorm:"index"`
	ReviewerID  string    `json:"reviewer_id"`
	Status      string    `json:"status"` // pending, approved, rejected
	Comment     string    `json:"comment"`
	CreatedAt   time.Time `json:"created_at"`
}

// BeforeCreate generates ULID before inserting
func (kr *KnowledgeReview) BeforeCreate(tx *gorm.DB) error {
	if kr.ID == "" {
		kr.ID = ulid.Make().String()
	}
	return nil
}

// TableName specifies the table name
func (KnowledgeReview) TableName() string {
	return "knowledge_reviews"
}

// ObsidianMapping stores Obsidian vault path mappings
type ObsidianMapping struct {
	ID         string     `json:"id" gorm:"primaryKey"`
	VaultPath  string     `json:"vault_path" gorm:"not null"`
	LocalPath  string     `json:"local_path" gorm:"not null"`
	CategoryID string     `json:"category_id"`
	AutoSync   bool       `json:"auto_sync" gorm:"default:false"`
	LastSyncAt *time.Time `json:"last_sync_at"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
}

// TableName specifies the table name
func (ObsidianMapping) TableName() string {
	return "obsidian_mappings"
}

File: apps/web/src/services/knowledge.ts
import { api, ApiResponse } from '@/utils/api'

export interface Category {
  id: string
  name: string
  description: string
  level: number
  sort_order: number
  children: Category[]
}

export interface Tag {
  id: string
  name: string
  color: string
  count: number
}

export interface Knowledge {
  id: string
  title: string
  content: string
  category_id: string
  author_id: string
  tags: Tag[]
  status: 'draft' | 'published' | 'archived'
  version: number
  view_count: number
  source: string
  created_at: string
  updated_at: string
  published_at?: string
}

export interface KnowledgeReview {
  id: string
  knowledge_id: string
  reviewer_id: string
  status: 'pending' | 'approved' | 'rejected'
  comment: string
  created_at: string
  knowledge?: Knowledge
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

export interface CreateCategoryRequest {
  name: string
  description?: string
  parent_id?: string
}

export interface UpdateCategoryRequest {
  name: string
  description?: string
}

export interface MoveCategoryRequest {
  parent_id?: string
}

export interface CreateKnowledgeRequest {
  title: string
  content: string
  category_id: string
  tag_ids?: string[]
  source?: string
}

export interface UpdateKnowledgeRequest {
  title: string
  content: string
  category_id?: string
  tag_ids?: string[]
}

export interface CreateTagRequest {
  name: string
  color?: string
}

export interface SubmitReviewRequest {
  reviewer_id: string
}

export interface ReviewActionRequest {
  comment?: string
}

// Category APIs
export const categoryApi = {
  getTree: () => api.get<ApiResponse<Category[]>>('/categories'),
  create: (data: CreateCategoryRequest) => api.post<ApiResponse<Category>>('/categories', data),
  update: (id: string, data: UpdateCategoryRequest) => api.put<ApiResponse<Category>>(`/categories/${id}`, data),
  delete: (id: string) => api.delete<ApiResponse<void>>(`/categories/${id}`),
  move: (id: string, data: MoveCategoryRequest) => api.post<ApiResponse<void>>(`/categories/${id}/move`, data),
}

// Knowledge APIs
export const knowledgeApi = {
  list: (params?: {
    page?: number
    page_size?: number
    category_id?: string
    status?: string
    tag_id?: string
    search?: string
  }) => api.get<ApiResponse<PaginatedResponse<Knowledge>>>('/knowledge', { params }),
  getById: (id: string) => api.get<ApiResponse<Knowledge>>(`/knowledge/${id}`),
  create: (data: CreateKnowledgeRequest) => api.post<ApiResponse<Knowledge>>('/knowledge', data),
  update: (id: string, data: UpdateKnowledgeRequest) => api.put<ApiResponse<Knowledge>>(`/knowledge/${id}`, data),
  delete: (id: string) => api.delete<ApiResponse<void>>(`/knowledge/${id}`),
  publish: (id: string) => api.post<ApiResponse<void>>(`/knowledge/${id}/publish`),
  archive: (id: string) => api.post<ApiResponse<void>>(`/knowledge/${id}/archive`),
  submitReview: (id: string, data: SubmitReviewRequest) => api.post<ApiResponse<KnowledgeReview>>(`/knowledge/${id}/review`, data),
}

// Tag APIs
export const tagApi = {
  list: () => api.get<ApiResponse<Tag[]>>('/tags'),
  create: (data: CreateTagRequest) => api.post<ApiResponse<Tag>>('/tags', data),
  delete: (id: string) => api.delete<ApiResponse<void>>(`/tags/${id}`),
}

// Review APIs
export const reviewApi = {
  getPending: () => api.get<ApiResponse<KnowledgeReview[]>>('/reviews/pending'),
  approve: (reviewId: string, data: ReviewActionRequest) => api.post<ApiResponse<void>>(`/reviews/${reviewId}/approve`, data),
  reject: (reviewId: string, data: ReviewActionRequest) => api.post<ApiResponse<void>>(`/reviews/${reviewId}/reject`, data),
}
